packages/tangle/messagemetadata.go

```diff
@@ -236,6 +242,38 @@ func (m *MessageMetadata) SetEligible(eligible bool) (modified bool) {
 	return
 }
 
+// IsInvalid returns true if the message represented by this metadata is invalid. False otherwise.
+func (m *MessageMetadata) IsInvalid() (result bool) {
+	m.invalidMutex.RLock()
+	result = m.invalid
+	m.invalidMutex.RUnlock()
+
+	return
+}
+
+// SetInvalid sets the message associated with this metadata as invalid.
+// It returns true if the invalid status is modified. False otherwise.
+func (m *MessageMetadata) SetInvalid(invalid bool) (modified bool) {
+	m.invalidMutex.RLock()
+	if m.invalid != invalid {
+		m.invalidMutex.RUnlock()
+
+		m.invalidMutex.Lock()
+		if m.invalid != invalid {
+			m.invalid = invalid
+			m.SetModified()
+
+			modified = true
+		}
+		m.invalidMutex.Unlock()
+
+	} else {
+		m.invalidMutex.RUnlock()
+	}
+
+	return
+}
```

Comment:
Please get rid of this ugly double locking business and add guard style. The code gets much harder to read and maintain even though it does such a simple thing. I agree that it might make sense to do this for heavy operations (e.g. computing a messageID) but setting a bool is definitely not one of them.

Something like this
```
lock()
defer unlock()

if m.invalid == invalid {
return false
}

m.invalid = invalid
m.SetModified()
return true

```

Reply:
Agree, I'll also refactor other bool setters.